### ISSUE TITLE
ansible-lint 25.9.1 fails on pre-commit with python 3.11

### ISSUE THREAD

**Issue by ylascombe** [ansible/ansible-lint#4812]
##### Summary

Since version `25.9.1` of ansible-lint, our pre-commit check fails with error `RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.13'`

In our setup, we systematically run `pre-commit autoupdate` before running `pre-commit run --all-files` and since the relase  `25.9.1` of ansible-lint we are getting this error.
I have tested with previous versions, it still works on `25.9.0` and start to fail on `25.9.1`

##### Issue Type

- Bug Report

##### OS / ENVIRONMENT

```console (paste below)
mise exec -- ansible-lint --version
ansible-lint 25.2.1 using ansible-core:2.19.1 ansible-compat:25.8.1 ruamel-yaml:0.18.15 ruamel-yaml-clib:0.2.12
```

- ansible installed with uvx (from mise)
- ansible-lint with uvx (from mise)

### Some details about the error (printed by [mise](https://github.com/jdx/mise)
Log of error:
```
### version information

` `  `
pre-commit version: 4.3.0
git --version: git version 2.43.0
sys.version:
    3.11.11 (main, Mar 17 2025, 21:02:09) [Clang 20.1.0 ]
sys.executable: /home/ylascombe/.cache/pre-commit-zipapp/ccbhuABpxV0HhQRu4I3g_LH0CXYFgf2Zd4Xlad7E818/python
os.name: posix
sys.platform: linux
` ` `

### error information

` ` `
An unexpected error has occurred: CalledProcessError: command: ('/home/ylascombe/.local/share/mise/installs/python/3.11.11/bin/python3', '/home/ylascombe/.cache/pre-commit-zipapp/ccbhuABpxV0HhQRu4I3g_LH0CXYFgf2Zd4Xlad7E818/python', '-mvirtualenv', '/home/ylascombe/.cache/pre-commit/repomcgopz1w/py_env-python3.13', '-p', 'python3.13')
return code: 1
stdout:
    RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.13'
stderr: (none)
` ` `

` ` `
Traceback (most recent call last):
  File "/home/ylascombe/.cache/pre-commit-zipapp/ccbhuABpxV0HhQRu4I3g_LH0CXYFgf2Zd4Xlad7E818/wheels/pre_commit-4.3.0-py2.py3-none-any.whl/pre_commit/error_handler.py", line 73, in error_handler
    yield
  File "/home/ylascombe/.cache/pre-commit-zipapp/ccbhuABpxV0HhQRu4I3g_LH0CXYFgf2Zd4Xlad7E818/wheels/pre_commit-4.3.0-py2.py3-none-any.whl/pre_commit/main.py", line 417, in main
    return run(args.config, store, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ylascombe/.cache/pre-commit-zipapp/ccbhuABpxV0HhQRu4I3g_LH0CXYFgf2Zd4Xlad7E818/wheels/pre_commit-4.3.0-py2.py3-none-any.whl/pre_commit/commands/run.py", line 442, in run
    install_hook_envs(to_install, store)
  File "/home/ylascombe/.cache/pre-commit-zipapp/ccbhuABpxV0HhQRu4I3g_LH0CXYFgf2Zd4Xlad7E818/wheels/pre_commit-4.3.0-py2.py3-none-any.whl/pre_commit/repository.py", line 229, in install_hook_envs
    _hook_install(hook)
  File "/home/ylascombe/.cache/pre-commit-zipapp/ccbhuABpxV0HhQRu4I3g_LH0CXYFgf2Zd4Xlad7E818/wheels/pre_commit-4.3.0-py2.py3-none-any.whl/pre_commit/repository.py", line 85, in _hook_install
    lang.install_environment(
  File "/home/ylascombe/.cache/pre-commit-zipapp/ccbhuABpxV0HhQRu4I3g_LH0CXYFgf2Zd4Xlad7E818/wheels/pre_commit-4.3.0-py2.py3-none-any.whl/pre_commit/languages/python.py", line 226, in install_environment
    cmd_output_b(*venv_cmd, cwd='/')
  File "/home/ylascombe/.cache/pre-commit-zipapp/ccbhuABpxV0HhQRu4I3g_LH0CXYFgf2Zd4Xlad7E818/wheels/pre_commit-4.3.0-py2.py3-none-any.whl/pre_commit/util.py", line 111, in cmd_output_b
    raise CalledProcessError(returncode, cmd, stdout_b, stderr_b)
pre_commit.util.CalledProcessError: command: ('/home/ylascombe/.local/share/mise/installs/python/3.11.11/bin/python3', '/home/ylascombe/.cache/pre-commit-zipapp/ccbhuABpxV0HhQRu4I3g_LH0CXYFgf2Zd4Xlad7E818/python', '-mvirtualenv', '/home/ylascombe/.cache/pre-commit/repomcgopz1w/py_env-python3.13', '-p', 'python3.13')
return code: 1
stdout:
    RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.13'
stderr: (none)
```

##### STEPS TO REPRODUCE

```
cat .pre-commit-config.yaml 
---
repos:
  - repo: https://github.com/ansible-community/ansible-lint.git
    rev: v25.9.1 # put latest release tag from https://github.com/ansible-community/ansible-lint/releases/
    hooks:
      - id: ansible-lint
        # Uncomment if you need the full Ansible community bundle instead of ansible-core:
        additional_dependencies:
          - ansible
          - distlib

# ...
```

```
pre-commit run --all-files

[INFO] Installing environment for https://github.com/ansible-community/ansible-lint.git.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('/home/ylascombe/.local/share/mise/installs/python/3.11.11/bin/python3', '/home/ylascombe/.cache/pre-commit-zipapp/ccbhuABpxV0HhQRu4I3g_LH0CXYFgf2Zd4Xlad7E818/python', '-mvirtualenv', '/home/ylascombe/.cache/pre-commit/repomcgopz1w/py_env-python3.13', '-p', 'python3.13')
return code: 1
stdout:
    RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.13'
stderr: (none)
Check the log at /home/ylascombe/.cache/pre-commit/pre-commit.log
```

[minimum complete verifiable example]: http://stackoverflow.com/help/mcve

**Comment by ssbarnea** [comment]
Sorry but from now on ansible-lint hook will have a hardcoded python version mentioned. This is because pre-commit cool does not have a way de declare a minimum version of python and on many systems the default is too old for picking the correct version of ansible-core.

If you do not have this python available, you can override the value in your hook definition.

**Comment by reinout** [comment]
I've followed [this advice](https://github.com/pre-commit/pre-commit/issues/1375#issuecomment-804399839) from a pre-commit issue to put this at the top of my `.pre-commit-config.yaml`:

    default_language_version:
      python: python3

In my github workflow I install a relatively new python (3.12, which is fine) as that fits my project best:

      - name: Set up Python
        uses: actions/setup-python@v5
        with:
          python-version: 3.12

But... that means pre-commit won't run unless I specifically set `language_version` for ansible-lint, as it requires explicitly 3.13. Or I have to install 3.13 into the github action, too. (I set `language_version: python3`, btw, so the same as the `default_language_version`).

The change *helps* people with old versions, but *might break workflows* for people that are nicely up-to-date. I wonder if there's an alternative solution, though I can't think of one right now.

**Comment by ssbarnea** [comment]
@reinout I feel you and this is why I reopen it, even if I do not know what path we should take. Soon we will start requiring py312 anyway. The reality is that is outside our power to fix pre-commit tool itself and the creator does not seem to add a fix for it. It is even worse because when using pre-commit.ci service, you cannot run setup-python to change the current python version.

**Comment by reinout** [comment]
I did a bit of searching on github and found https://github.com/adamchainz/blacken-docs/pull/258 and especially https://github.com/psf/black/pull/2430 .

The `pyproject.toml` specifies `requires-python = ">=3.10"`, so on older pythons, I'd expect a reasonably okay error message. Though I might be mistaken, as old python versions are perhaps too old to understand the `requires-python` instruction. 

I've looked in the issues, but couldn't find a report showing the problem on old python versions. Is there one? Perhaps I spot something that can help.

I had an idea for a fix, but that didn't work out. Perhaps `minimum_pre_commit_version` could specify a pre-commit version that itself depended on a new-enough python version. Problem: pre-commit specifies 3.9 as the minimum version :-) So alas.

**Comment by ylascombe** [comment]
Thanks @ssbarnea and @reinout for your answers. 

Thanks yo your link @reinout , I have added two attribute in my pre-commit hook config and now it is working again:
```
  - repo: https://github.com/ansible-community/ansible-lint.git
    rev: v25.9.2 # put latest release tag from https://github.com/ansible-community/ansible-lint/releases/
    hooks:
      - id: ansible-lint
        # Uncomment if you need the full Ansible community bundle instead of ansible-core:
        language_version: python3.11   # First change, force python version
        additional_dependencies:
          - ansible-core==2.16.14   # second change, lock my ansible-core version
          - distlib
```

For me we can close this issue.

Thanks

**Comment by KSmanis** [comment]
For anyone interested, [prek](https://github.com/j178/prek) is capable of automatically provisioning Python toolchains, i.e., it doesn't matter what Python version you are running (locally or on CI), prek will automatically provision the missing Python version for ansible-lint.
